### PR TITLE
add spades 3.15.4

### LINKED ISSUE #255: update SPAdes dockerfile with Ubuntu 18 (20?) to catch WSL errors
Need to upgrade the base image to either ubuntu bionic or focal, `apt install python3 python3-distutils`

This will catch anyone trying to run the SPAdes linux binaries on WSL2 and spit out an informative error message

more info here: https://github.com/ablab/spades/issues/838

### PR DESCRIPTION
**Pull request by kapsakcj**
- [x] This comment contains a description of what is in the pull request.

This PR adds a dockerfile for spades 3.15.4, now with `ubuntu:focal` as the base, and only python 3.8.10 (no python 2.x.x). This was done to fix #255 and allow for WSL2 users to see an informative error message saying that WSL2 is not supported....but when I tested on WSL2, `spades.py --test` ran successfully (despite me expecting it to fail). IDK. Going to rejoice in the fact that it runs (or at least the test runs) in WSL2 👍 

Otherwise on Linux this docker image runs successfully as expected:
```
curtis_kapsak@curtis-dev-ubuntu20:~$ docker build -t kapsakcj/spades:3.15.4 spades/3.15.4/
Sending build context to Docker daemon  3.072kB
Step 1/15 : FROM ubuntu:focal
 ---> d13c942271d6
Step 2/15 : ARG SPAdesVer=3.15.4
 ---> Using cache
 ---> 65105a571272

[...]

Successfully built 7875e79d0715
Successfully tagged kapsakcj/spades:3.15.4

curtis_kapsak@curtis-dev-ubuntu20:~$ docker_run -it kapsakcj/spades:3.15.4
groups: cannot find name for group ID 1002
I have no name!@d31d3d22961a:/data$ python --version
Python 3.8.10
I have no name!@d31d3d22961a:/data$ python3 --version
Python 3.8.10
I have no name!@d31d3d22961a:/data$ spades.py --test


== Warning ==  No assembly mode was specified! If you intend to assemble high-coverage multi-cell/isolate data, use '--isolate' option.


Command line: /SPAdes-3.15.4-Linux/bin/spades.py        --test

System information:
  SPAdes version: 3.15.4
  Python version: 3.8.10
  OS: Linux-5.11.0-1029-gcp-x86_64-with-glibc2.29

[...]

===== Check test started.


== Running: /usr/bin/python /SPAdes-3.15.4-Linux/share/spades/spades_pipeline/scripts/check_test_script.py --mode common --result_contigs_filename /data/spades_test/contigs.fasta --result_scaffolds_filename /data/spades_test/scaffolds.fasta

========= TEST PASSED CORRECTLY.

===== Check test finished.


===== Breaking scaffolds started.


== Running: /usr/bin/python /SPAdes-3.15.4-Linux/share/spades/spades_pipeline/scripts/breaking_scaffolds_script.py --result_scaffolds_filename /data/spades_test/scaffolds.fasta --misc_dir /data/spades_test/misc --threshold_for_breaking_scaffolds 3


===== Breaking scaffolds finished.


===== Terminate started.


===== Terminate finished.

 * Corrected reads are in /data/spades_test/corrected/
 * Assembled contigs are in /data/spades_test/contigs.fasta
 * Assembled scaffolds are in /data/spades_test/scaffolds.fasta
 * Paths in the assembly graph corresponding to the contigs are in /data/spades_test/contigs.paths
 * Paths in the assembly graph corresponding to the scaffolds are in /data/spades_test/scaffolds.paths
 * Assembly graph is in /data/spades_test/assembly_graph.fastg
 * Assembly graph in GFA format is in /data/spades_test/assembly_graph_with_scaffolds.gfa

======= SPAdes pipeline finished WITH WARNINGS!

=== Error correction and assembling warnings:
 * 0:00:01.065     1M / 155M  WARN    General                 (launcher.cpp              : 178)   Your data seems to have high uniform coverage depth. It is strongly recommended to use --isolate option.
======= Warnings saved to /data/spades_test/warnings.log

SPAdes log can be found here: /data/spades_test/spades.log

Thank you for using SPAdes!
```

- [x] Build your own docker image using a Dockerfile
  - [x] Directory structure should be name of the tool in lower case with special characters removed with a subdirectory of the version number (i.e. spades/3.12.0/Dockerfile)
  - [x] Includes the recommended LABELS
- [x] (Optional) Dockerfile is built with best practices and has been approved by a linter (such as https://hadolint.github.io/hadolint/)
- [x] Edit main README.md
- [x] Ensure tool is listed in Program_Licenses.md
- [x] Create a simple container-specific README.md in the same directory as the Dockerfile (i.e. spades/3.12.0/README.md)
- [x] Update GitHub actions workflow if needed
- [x] Any files required for building are located in the same directory as the Dockerfile (i.e. spades/3.12.0/my_spades_tests.sh)
- [x] Have successfully run the workflow "Test <program name> image" in your forked repository

### CODE REVIEW

File: spades/3.15.4/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,34 @@
+FROM ubuntu:focal
```
Comment (erinyoung) [inline spades/3.15.4/Dockerfile:1]: I'm not going to lie. I'm going to grab this file and see if I can create a test for it. 

Thank you for being an involuntary guinea pig.

Comment (kapsakcj) [reply spades/3.15.4/Dockerfile:1]: go right ahead 😁

**Comment by erinyoung** [comment]
I know I'm not Kevin, but I've reviewed your file.

1) it builds and 2) it works

BUT! I have some suggestions for your dockerfile to incorporate github actions. It'll make this 10000% cooler.

This is how I adjusted your file and then I built this with `--target test`. Basically I added a 'as app' on the FROM ubuntu:focal and a test at the end.

```
$ cat spades/3.15.4/Dockerfile 
FROM ubuntu:focal as app

ARG SPAdesVer="3.15.4"

LABEL base.image="ubuntu:focal"
LABEL dockerfile.version="1"
LABEL software="SPAdes"
LABEL software.version="3.15.4"
LABEL description="de novo DBG genome assembler"
LABEL website="https://github.com/ablab/spades"
LABEL license="https://github.com/ablab/spades/blob/v3.15.3/assembler/LICENSE"
LABEL maintainer="Curtis Kapsak"
LABEL maintainer.email="kapsakcj@gmail.com"

# install dependencies; cleanup apt garbage
# python v3.8.10 is installed here; point 'python' to python3
RUN apt-get update && apt-get install --no-install-recommends -y python3 \
 python3-distutils \
 wget && \
 apt-get autoclean && rm -rf /var/lib/apt/lists/* && \
 update-alternatives --install /usr/bin/python python /usr/bin/python3 10

# install SPAdes binary; make /data
RUN wget http://cab.spbu.ru/files/release${SPAdesVer}/SPAdes-${SPAdesVer}-Linux.tar.gz && \
  tar -xzf SPAdes-${SPAdesVer}-Linux.tar.gz && \
  rm -r SPAdes-${SPAdesVer}-Linux.tar.gz && \
  mkdir /data

# set PATH and locale settings for singularity
ENV LC_ALL=C.UTF-8 \
    PATH="${PATH}:/SPAdes-${SPAdesVer}-Linux/bin"

WORKDIR /data

# testing
FROM app as test

RUN spades.py --test
```
Then I built the container with 
```
docker build --target test --tag spades:test2 spades/3.15.4
```

Then you'll also want a corresponding test file in `.github/workflows`

```
$ cat .github/workflows/test-spades.yml 
name: Test spades image

on:
  workflow_dispatch:
  pull_request:
    paths:
      - 'spades/3.15.4/Dockerfile'

jobs:
  build-to-test:
    uses: ./.github/workflows/build-to-test.yml
    with:
      path_to_context: "spades/3.15.4"
      dockerfile_name: "Dockerfile"
      cache: "spades"
```

**Comment by kapsakcj** [comment]
Converted this PR to a draft so that I can add GH actions workflows for testing and deploying the docker image. Will remove draft state when it's ready for review

Want to get some feedback on my GH actions workflows for `pangolin` and then I will add @erinyoung 's suggested changes with the feedback in mind.

I think these workflows will really reduce the hands on time for these docker image updates & additions 👍

**Comment by kapsakcj** [comment]
OK I've added a test YML as well as a deploy YML. Also updated the docker image as per your suggestions.

Ready for review

**Review by erinyoung — APPROVED** [review]
Looks good!

**Pull request merged**